src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorOptimizedTransform.java

```diff
@@ -0,0 +1,155 @@
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.dataflow.sdk.values.PCollectionView;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMSequenceDictionary;
+import org.broadinstitute.hellbender.engine.dataflow.DoFnWLog;
+import org.broadinstitute.hellbender.engine.dataflow.transforms.composite.AddContextDataToReadOptimized;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.tools.recalibration.RecalibrationTables;
+import org.broadinstitute.hellbender.tools.recalibration.covariates.StandardCovariateList;
+import org.broadinstitute.hellbender.tools.walkers.bqsr.RecalibrationEngine;
+import org.broadinstitute.hellbender.utils.io.IOUtils;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+
+/**
+ * Base Quality Score Recalibration, phase 1
+ * This contains the Dataflow-specific bits. Most of the work is done by BaseRecalibratorFn.
+ * The Hellbender-specific bits are in BaseRecalibratorDataflow2.
+ *
+ * Unlike the non-optimized version, this one works on a whole shard at a time.
+ */
+public final class BaseRecalibratorOptimizedTransform extends PTransform<PCollection<AddContextDataToReadOptimized.ContextShard>, PCollection<RecalibrationTables>> {
+    private static final long serialVersionUID = 1L;
+    private PCollectionView<SAMFileHeader> headerView;
+    private PCollectionView<SAMSequenceDictionary> refDictionary;
+    BaseRecalibrationArgumentCollection BRAC;
+
+    public BaseRecalibratorOptimizedTransform(final PCollectionView<SAMFileHeader> headerView, PCollectionView<SAMSequenceDictionary> refDictionary, BaseRecalibrationArgumentCollection BRAC) {
+        this.headerView = headerView;
+        this.refDictionary = refDictionary;
+        this.BRAC = BRAC;
+    }
+
+
+    @Override
+    public PCollection<RecalibrationTables> apply( PCollection<AddContextDataToReadOptimized.ContextShard> input ) {
+        PCollection<RecalibrationTables> oneStatPerWorker =
+            input.apply(ParDo.named("BaseRecalibrator")
+                .withSideInputs(headerView, refDictionary)
+                .of(new BaseRecalibratorOptimizedFn(headerView, refDictionary, BRAC)));
+        /*
+        // TEMPORARY: for debugging, save a copy of the RT.
+        oneStatPerWorker.apply(ParDo.of(new DoFnWLog<RecalibrationTables, Void>("saveRT") {
+            private static final long serialVersionUID = 1L;
+            @Override
+            public void processElement(ProcessContext c) throws Exception {
+                int randomName = Math.abs(new Random().nextInt());
+                try (ObjectOutputStream oos = new ObjectOutputStream(BucketUtils.createFile("gs://jpmartin/test-output/recaltable-"+randomName+".sj", c.getPipelineOptions()))) {
+                    oos.writeObject(c.element());
+                    oos.flush();
+                }
+            }
+        }));
+        */
```

Comment:
Are you sure? As per instructions, the code still isn't correct. We may need this code back very soon.